test/testsuite.h

```diff
@@ -53,6 +55,28 @@ class TestFixture : public ErrorLogger {
 
     bool assert_(const char * const filename, const unsigned int linenr, const bool condition) const;
 
+    template <typename T, typename U>
+    bool assertEquals(const char* const filename, const unsigned int linenr, const T& expected, const U& actual, const std::string& msg = emptyString) const {
```

Comment:
I don't see what this `assertEquals` accomplish that we don't have with the overloads.

It is a good idea to put code in source file I think we should be careful about this.

Reply:
I think putting it in the header is effectively unavoidable if it is a template. You're right in your comment below that a significant portion that is not dependent on the template parameters could be refactored out and put back in the cpp file.

The template doesn't have huge value here. It exists to expose the warnings that I was encountering in #3056 so I can fix them preemptively rather than further complicate the scope of that PR. The specific use case I had in that PR was for asserting equality of two pointers but I think more broadly 
- it makes sense to be able to assert that things which can be compared for equality (and converted to strings for the purpose of failure messages) are equal
- and it makes sense to define equality based on the types involved.

For that first point the pointer thing is a good example. Although it's not ideal to generate an error message based on pointer values because they're not sure useful they can be a little bit useful (especially when null) and sometimes the assertion itself is more important than the error message.

For the second point consider ```ASSERT_EQUALS('a', 'b');```. Before my change  this gives
```
Expected:
97
Actual:
98
```
and after it gives
```
Expected:
a
Actual:
b
```
which I think is preferable.

Now consider ```ASSERT_EQUALS(0.0, 0.1)```. This actually passes before my changes because the 0.1 is implicitly truncated into a long long. Of course there is a ASSERT_EQUALS_DOUBLE which you would prefer in many circumstances but in some few circumstances you don't need the tolerance.